### PR TITLE
Twilio SDK Instrumentation

### PR DESCRIPTION
**Pull request by darylrobbins**
Improved instrumentation for the [Twilio Java SDK](https://github.com/twilio/twilio-java)

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -82,6 +82,7 @@ include ':dd-java-agent:instrumentation:spring-web'
 include ':dd-java-agent:instrumentation:spring-webflux'
 include ':dd-java-agent:instrumentation:spymemcached-2.12'
 include ':dd-java-agent:instrumentation:tomcat-classloading'
+include ':dd-java-agent:instrumentation:twilio'
```
Comment (tylerbenson) [inline settings.gradle]: nit... this should be alphabetical.  Also, depending on the versions of the client this instrumentation works for, the name should include the earliest compatible version.

Comment (tylerbenson) [reply settings.gradle]: You also need to add a `twilio.gradle` file in the instrumentation project.  Prob easiest to copy the `apache-http-client` one (the AWS one is a bit more complicated) and modify to match this project.

File: dd-java-agent/instrumentation/twilio/twilio.gradle [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  compile project(':dd-java-agent:agent-tooling')
+
+  compile deps.bytebuddy
+  compile deps.opentracing
+  annotationProcessor deps.autoservice
+  implementation deps.autoservice
+
+  testCompile group: 'com.twilio.sdk', name: 'twilio', version: '7.36.2'
+  testCompile project(':dd-java-agent:testing')
+  testCompile project(':dd-java-agent:instrumentation:java-concurrent')
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/twilio/twilio.gradle]: Also add a compile dependency on the http client instrumentation project so that instrumentation also shows up in your tests.

File: dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioClientDecorator.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  /** Decorate trace based on service execution metadata */
+  public Span onServiceExecution(
+      final Span span, final Object serviceExecutor, final String methodName) {
+
+    // Drop common package prefix (com.twilio.rest)
+    final String simpleClassName =
+        serviceExecutor.getClass().getCanonicalName().replaceFirst("^com\\.twilio\\.rest\\.", "");
+
+    span.setTag(DDTags.RESOURCE_NAME, String.format("%s.%s", simpleClassName, methodName));
```
[the remark this answers is not in the document]
Comment (darylrobbins) [reply dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioClientDecorator.java]: Actually, I implemented this myself for a reason. I have included the minimum from the qualified name of the class that is required to uniquely identify it. For example, `com.twilio.rest.chat.v1.ServiceCreator`, `com.twilio.rest.chat.v2.ServiceCreator`, and `com.twilio.rest.ipmessaging.v1.ServiceCreator` are all different classes.

File: dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  public static class TwilioClientAdvice {
+
+    /** Method entry instrumentation */
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static Scope methodEnter(
+        @Advice.This final Object that, @Advice.Origin("#m") final String methodName) {
+
+      System.out.println("START");
```
[the remark this answers is not in the document]
Comment (darylrobbins) [reply dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java]: Yep, this was a test that I introduced in parallel of creating the PR. Removed.

File: settings.gradle (hunk 1 above)
Comment (darylrobbins) [reply settings.gradle]: Opps, was off by one in terms of alphabetization. Fixed. Also, I had already created my twilio.gradle, but forgot to add it when I made my git commit.

File: dd-java-agent/instrumentation/twilio/twilio.gradle (hunk 2 above)
Comment (darylrobbins) [reply dd-java-agent/instrumentation/twilio/twilio.gradle]: Not needed, since I'm mocking the TwilioRestClient and not calling the HTTP Client at all.

File: dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+      final int callDepth = CallDepthThreadLocalMap.incrementCallDepth(Twilio.class);
+      if (callDepth > 0) {
+        return null;
+      }
+
+      final Tracer tracer = GlobalTracer.get();
+      final Scope scope = tracer.buildSpan("twilio.sdk").startActive(true);
```
[the remark this answers is not in the document]
Comment (darylrobbins) [reply dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java]: Corrected, but this seemed to break all of the tests.

File: dd-java-agent/instrumentation/twilio/twilio.gradle (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/twilio/twilio.gradle]: Ah, ok... would it be possible to have at least one test that verifies it with the real http client?  I think this is important because it would verify the expected real world behavior.

File: dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+          DECORATE.onError(span, throwable);
+          DECORATE.beforeFinish(span);
+
+          // If we're calling an async operation, we still need to finish the span when it's
+          // complete; set an appropriate callback
+          if (isAsync && response instanceof ListenableFuture) {
+            Futures.addCallback(
+                (ListenableFuture) response,
+                new SpanFinishingCallback(span),
+                Twilio.getExecutorService());
```
[the remark this answers is not in the document]
Comment (darylrobbins) [reply dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioInstrumentation.java]: It can't be done in OnMethodEnter, since the Future is returned by the method.

File: dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioClientDecorator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (labbati) [reply dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioClientDecorator.java]: A couple of thoughts here:
- I like the idea of keeping the resource name to the minimum possible in order to focus on the relevant part
- I am not a super fan of completely removing the info (in general) because `chat.v1.ServiceCreator` is a perfectly valid package name and this may generate confusion in who is not super familiar with the specific library

What if we take inspiration from LogBack's [ConversionWord](https://logback.qos.ch/manual/layouts.html#conversionWord) idea simplifying it?

Basically  in `BaseDecorator` we add n overload that abbreviates the package names to a one character dotted name, up to a configurable point.

E.g.
```
BaseDecorator.spanNameForMethod("com.twilio.rest.chat.v1.ServiceCreator", "someMethod", "com.twilio.rest" );

// "c.t.r.chat.v1.ServiceCreator"
```

Do you thing resource names would be to ugly to look at? 😄

File: dd-java-agent/instrumentation/twilio/twilio.gradle (hunk 2 above)
Comment (labbati) [reply dd-java-agent/instrumentation/twilio/twilio.gradle]: Is this ^^ still pending?

**Comment by labbati** [comment]
Thanks a lot @darylrobbins for working on this 🙇
